File: src/fr/adrienbrault/idea/symfony2plugin/ui/TwigSettingsForm.java
package fr.adrienbrault.idea.symfony2plugin.ui;

import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.options.Configurable;
import com.intellij.openapi.options.ConfigurationException;
import com.intellij.openapi.project.Project;
import com.intellij.ui.AnActionButton;
import com.intellij.ui.AnActionButtonRunnable;
import com.intellij.ui.AnActionButtonUpdater;
import com.intellij.ui.ToolbarDecorator;
import com.intellij.ui.table.TableView;
import com.intellij.util.ui.ColumnInfo;
import com.intellij.util.ui.ElementProducer;
import com.intellij.util.ui.ListTableModel;
import fr.adrienbrault.idea.symfony2plugin.Settings;
import fr.adrienbrault.idea.symfony2plugin.TwigHelper;
import fr.adrienbrault.idea.symfony2plugin.templating.path.TwigNamespaceSetting;
import fr.adrienbrault.idea.symfony2plugin.templating.path.TwigPath;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.event.TableModelEvent;
import javax.swing.event.TableModelListener;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class TwigSettingsForm implements Configurable {

    private JPanel panel1;
    private JPanel panelTableView;
    private JButton resetToDefault;
    private TableView<TwigPath> tableView;
    private Project project;
    private boolean changed = false;
    private ListTableModel<TwigPath> modelList;

    public TwigSettingsForm(@NotNull Project project) {

        this.project = project;

        this.tableView = new TableView<TwigPath>();
        this.modelList = new ListTableModel<TwigPath>(
            new NamespaceColumn(),
            new PathColumn(project),
            new TypeColumn(),
            new CustomColumn(),
            new DisableColumn()
        );

        this.attachItems(true);

        this.tableView.setModelAndUpdateColumns(this.modelList);

        this.modelList.addTableModelListener(new TableModelListener() {
            @Override
            public void tableChanged(TableModelEvent e) {
                TwigSettingsForm.this.changed = true;
            }
        });

        resetToDefault.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                super.mouseClicked(e);
                TwigSettingsForm.this.resetList();

                List<TwigPath> sortableLookupItems = new ArrayList<TwigPath>();
                sortableLookupItems.addAll(TwigHelper.getTwigNamespaces(TwigSettingsForm.this.project, false));
                Collections.sort(sortableLookupItems);

                for (TwigPath twigPath : sortableLookupItems) {
                    // dont use managed class here
                    // @TODO state to enabled (should not be here)
                    TwigSettingsForm.this.modelList.addRow(twigPath.clone().setEnabled(true));
                }
            }
        });
    }

    private void attachItems(boolean includeSettings) {

        // dont load on project less context
        if(this.project == null) {
            return;
        }
        
        List<TwigPath> sortableLookupItems = new ArrayList<TwigPath>();
        sortableLookupItems.addAll(TwigHelper.getTwigNamespaces(this.project, includeSettings));
        Collections.sort(sortableLookupItems);

        for (TwigPath twigPath : sortableLookupItems) {
            // dont use managed class here
            this.modelList.addRow(twigPath.clone());
        }
    }

    @Nls
    @Override
    public String getDisplayName() {
        return "Twig";
    }

    @Nullable
    @Override
    public String getHelpTopic() {
        return null;
    }

    @Nullable
    @Override
    public JComponent createComponent() {
        ToolbarDecorator tablePanel = ToolbarDecorator.createDecorator(this.tableView, new ElementProducer<TwigPath>() {
            @Override
            public TwigPath createElement() {
                //IdeFocusManager.getInstance(TwigSettingsForm.this.project).requestFocus(TwigNamespaceDialog.getWindows(), true);
                return null;  //To change body of implemented methods use File | Settings | File Templates.
            }

            @Override
            public boolean canCreateElement() {
                return true;  //To change body of implemented methods use File | Settings | File Templates.
            }
        });

        tablePanel.setEditAction(new AnActionButtonRunnable() {
            @Override
            public void run(AnActionButton anActionButton) {
                TwigSettingsForm.this.openTwigPathDialog(TwigSettingsForm.this.tableView.getSelectedObject());
            }
        });


        tablePanel.setAddAction(new AnActionButtonRunnable() {
            @Override
            public void run(AnActionButton anActionButton) {
               TwigSettingsForm.this.openTwigPathDialog(null);
            }
        });

        tablePanel.setEditActionUpdater(new AnActionButtonUpdater() {
            @Override
            public boolean isEnabled(AnActionEvent e) {
                TwigPath twigPath = TwigSettingsForm.this.tableView.getSelectedObject();
                return twigPath != null && twigPath.isCustomPath();
            }
        });

        tablePanel.setRemoveActionUpdater(new AnActionButtonUpdater() {
            @Override
            public boolean isEnabled(AnActionEvent e) {
                TwigPath twigPath = TwigSettingsForm.this.tableView.getSelectedObject();
                return twigPath != null && twigPath.isCustomPath();
            }
        });

        tablePanel.disableUpAction();
        tablePanel.disableDownAction();

        this.panelTableView.add(tablePanel.createPanel());

        return this.panel1;
    }

    @Override
    public boolean isModified() {
        return this.changed;
    }

    @Override
    public void apply() throws ConfigurationException {
        List<TwigNamespaceSetting> twigPaths = new ArrayList<TwigNamespaceSetting>();

        for(TwigPath twigPath :this.tableView.getListTableModel().getItems()) {
            // only custom and disabled path need to save
            if((!twigPath.isEnabled() && twigPath.getRelativePath(this.project) != null) || twigPath.isCustomPath()) {
                twigPaths.add(new TwigNamespaceSetting(twigPath.getNamespace(), twigPath.getRelativePath(this.project), twigPath.isEnabled(), twigPath.getNamespaceType(), twigPath.isCustomPath()));
            }
        }

        getSettings().twigNamespaces = twigPaths;
        this.changed = false;
    }

    private Settings getSettings() {
        return Settings.getInstance(this.project);
    }

    private void resetList() {
        // clear list, easier?
        while(this.modelList.getRowCount() > 0) {
            this.modelList.removeRow(0);
        }

    }

    @Override
    public void reset() {
        this.resetList();
        this.attachItems(true);
        this.changed = false;
    }

    @Override
    public void disposeUIResources() {
        this.resetList();
    }

    private class NamespaceColumn extends ColumnInfo<TwigPath, String> {

        public NamespaceColumn() {
            super("Namespace");
        }

        @Nullable
        @Override
        public String valueOf(TwigPath twigPath) {
             return twigPath.getNamespace();
        }
    }

    private class PathColumn extends ColumnInfo<TwigPath, String> {

        private Project project;

        public PathColumn(Project project) {
            super("Path");
            this.project = project;
        }

        @Nullable
        @Override
        public String valueOf(TwigPath twigPath) {
            return twigPath.getRelativePath(this.project);
        }
    }

    private class CustomColumn extends ColumnInfo<TwigPath, String> {

        public CustomColumn() {
            super("Parser");
        }

        @Nullable
        @Override
        public String valueOf(TwigPath twigPath) {
            return twigPath.isCustomPath() ? "Custom" : "Internal";
        }
    }

    private class TypeColumn extends ColumnInfo<TwigPath, String> {

        public TypeColumn() {
            super("Type");
        }

        @Nullable
        @Override
        public String valueOf(TwigPath twigPath) {
            return twigPath.getNamespaceType().toString();
        }
    }

    private abstract class BooleanColumn extends ColumnInfo<TwigPath, Boolean>
    {
        public BooleanColumn(String name) {
            super(name);
        }

        public boolean isCellEditable(TwigPath groupItem)
        {
            return true;
        }

        public Class getColumnClass()
        {
            return Boolean.class;
        }
    }

    private class DisableColumn extends BooleanColumn {

        public DisableColumn() {
            super("on");
        }

        public Boolean valueOf(TwigPath twigPath) {
            return twigPath.isEnabled();
        }

        public void setValue(TwigPath twigPath, Boolean value){
            twigPath.setEnabled(value);
            TwigSettingsForm.this.tableView.getListTableModel().fireTableDataChanged();
        }

        public int getWidth(JTable table) {
            return 50;
        }

    }

    private void openTwigPathDialog(@Nullable TwigPath twigPath) {
        TwigNamespaceDialog twigNamespaceDialog;
        if(twigPath == null) {
            twigNamespaceDialog = new TwigNamespaceDialog(project, this.tableView);
        } else {
            twigNamespaceDialog = new TwigNamespaceDialog(project, this.tableView, twigPath);
        }

        Dimension dim = new Dimension();
        dim.setSize(500, 190);
        twigNamespaceDialog.setTitle("Twig Namespace");
        twigNamespaceDialog.setMinimumSize(dim);
        twigNamespaceDialog.pack();
        twigNamespaceDialog.setLocationRelativeTo(TwigSettingsForm.this.panel1);

        twigNamespaceDialog.setVisible(true);
    }

}

File: src/main/kotlin/fr/adrienbrault/idea/symfony2plugin/ui/TwigSettingsForm.kt
package fr.adrienbrault.idea.symfony2plugin.ui

import com.intellij.openapi.application.ModalityState
import com.intellij.openapi.application.ReadAction
import com.intellij.openapi.options.Configurable
import com.intellij.openapi.project.DumbService
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.io.FileUtil
import com.intellij.ui.ToolbarDecorator
import com.intellij.ui.table.TableView
import com.intellij.util.concurrency.AppExecutorUtil
import com.intellij.util.ui.ColumnInfo
import com.intellij.util.ui.ElementProducer
import com.intellij.util.ui.ListTableModel
import fr.adrienbrault.idea.symfony2plugin.Settings
import fr.adrienbrault.idea.symfony2plugin.templating.path.TwigNamespaceSetting
import fr.adrienbrault.idea.symfony2plugin.templating.path.TwigPath
import fr.adrienbrault.idea.symfony2plugin.templating.util.TwigUtil
import java.awt.BorderLayout
import java.awt.Dimension
import java.awt.FlowLayout
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.util.concurrent.atomic.AtomicInteger
import javax.swing.JButton
import javax.swing.JCheckBox
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JTable

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
class TwigSettingsForm(private val project: Project) : Configurable {
    private lateinit var panel1: JPanel
    private lateinit var chkTwigBundleNamespaceSupport: JCheckBox
    private lateinit var tableView: TableView<TwigPath>
    private var changed = false
    private lateinit var modelList: ListTableModel<TwigPath>
    private val reloadRequestId = AtomicInteger()

    @Volatile
    private var disposed = false

    private fun attachItems() {
        reloadTwigPaths(true, false)
    }

    private fun reloadTwigPaths(includeSettings: Boolean, forceEnableAll: Boolean) {
        if (project.isDefault) {
            return
        }

        resetList()

        if (DumbService.getInstance(project).isDumb) {
            tableView.emptyText.text = "Loading after indexing completes..."
        } else {
            tableView.emptyText.text = "Loading Twig namespaces..."
        }

        val requestId = reloadRequestId.incrementAndGet()

        ReadAction
            .nonBlocking<List<TwigPath>> {
                val sortableLookupItems = ArrayList(TwigUtil.getTwigNamespaces(project, includeSettings))
                sortableLookupItems.sortWith(TwigUtil.TwigPathNamespaceComparator())
                sortableLookupItems
            }
            .inSmartMode(project)
            .expireWhen { disposed || project.isDisposed || requestId != reloadRequestId.get() }
            .finishOnUiThread(ModalityState.any()) { applyTwigPaths(requestId, it, forceEnableAll) }
            .submit(AppExecutorUtil.getAppExecutorService())
    }

    private fun applyTwigPaths(requestId: Int, twigPaths: List<TwigPath>, forceEnableAll: Boolean) {
        if (disposed || project.isDisposed || requestId != reloadRequestId.get()) {
            return
        }

        resetList()

        twigPaths.forEach {
            modelList.addRow(if (forceEnableAll) TwigPath.createClone(it, true) else TwigPath.createClone(it))
        }

        if (twigPaths.isEmpty()) {
            tableView.emptyText.text = "No Twig namespaces found."
        }
    }

    override fun getDisplayName(): String = "Twig"

    override fun getHelpTopic(): String? = null

    override fun createComponent(): JPanel {
        disposed = false
        panel1 = JPanel(BorderLayout())
        val panelTableView = JPanel(BorderLayout())
        panel1.add(panelTableView, BorderLayout.CENTER)

        chkTwigBundleNamespaceSupport = JCheckBox("Support Bundle Namespaces")
        chkTwigBundleNamespaceSupport.isSelected = true
        chkTwigBundleNamespaceSupport.toolTipText = "Example: Foobar:Bar:Foo.html.twig (for older Symfony Versions)"
        val buttonJsonExample = JButton("JSON Example")
        val resetToDefault = JButton("Reset To Default")

        val northPanel = JPanel(FlowLayout(FlowLayout.LEFT))
        northPanel.add(JLabel("Manage Twig Namespaces"))
        northPanel.add(chkTwigBundleNamespaceSupport)
        northPanel.add(buttonJsonExample)
        northPanel.add(resetToDefault)
        panel1.add(northPanel, BorderLayout.NORTH)

        tableView = TableView()
        modelList = ListTableModel(
            NamespaceColumn(),
            PathColumn(project),
            TypeColumn(),
            CustomColumn(),
            DisableColumn(),
        )

        attachItems()

        tableView.setModelAndUpdateColumns(modelList)
        modelList.addTableModelListener { changed = true }

        resetToDefault.addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(e: MouseEvent) {
                super.mouseClicked(e)
                reloadTwigPaths(false, true)
            }
        })

        val tablePanel = ToolbarDecorator.createDecorator(tableView, object : ElementProducer<TwigPath> {
            override fun createElement(): TwigPath? {
                //IdeFocusManager.getInstance(TwigSettingsForm.this.project).requestFocus(TwigNamespaceDialog.getWindows(), true);
                return null //To change body of implemented methods use File | Settings | File Templates.
            }

            override fun canCreateElement(): Boolean {
                return true //To change body of implemented methods use File | Settings | File Templates.
            }
        })

        tablePanel.setEditAction { openTwigPathDialog(tableView.selectedObject) }
        tablePanel.setAddAction { openTwigPathDialog(null) }

        tablePanel.setEditActionUpdater {
            tableView.selectedObject?.isCustomPath == true
        }

        tablePanel.setRemoveActionUpdater {
            tableView.selectedObject?.isCustomPath == true
        }

        tablePanel.disableUpAction()
        tablePanel.disableDownAction()

        panelTableView.add(tablePanel.createPanel())

        buttonJsonExample.addActionListener { TwigJsonExampleDialog.open(panel1) }

        return panel1
    }

    override fun isModified(): Boolean {
        return changed || settings.twigBundleNamespaceSupport != chkTwigBundleNamespaceSupport.isSelected
    }

    override fun apply() {
        val twigPaths = ArrayList<TwigNamespaceSetting>()

        for (twigPath in tableView.listTableModel.items) {
            // Skip absolute paths - only relative paths are supported
            if (twigPath.isCustomPath && FileUtil.isAbsolute(twigPath.path)) {
                continue
            }

            val relativePath = twigPath.getRelativePath(project)

            // only custom and disabled path need to save
            if ((!twigPath.isEnabled && relativePath != null) || twigPath.isCustomPath) {
                twigPaths.add(
                    TwigNamespaceSetting(
                        twigPath.namespace,
                        relativePath,
                        twigPath.isEnabled,
                        twigPath.namespaceType,
                        twigPath.isCustomPath,
                    ),
                )
            }
        }

        settings.twigBundleNamespaceSupport = chkTwigBundleNamespaceSupport.isSelected
        settings.twigNamespaces = twigPaths
        changed = false
    }

    private val settings: Settings
        get() = Settings.getInstance(project)

    private fun resetList() {
        // clear list, easier?
        while (modelList.rowCount > 0) {
            modelList.removeRow(0)
        }
    }

    override fun reset() {
        attachItems()
        updateUIFromSettings()
        changed = false
    }

    private fun updateUIFromSettings() {
        chkTwigBundleNamespaceSupport.isSelected = settings.twigBundleNamespaceSupport
    }

    override fun disposeUIResources() {
        disposed = true
        reloadRequestId.incrementAndGet()
        resetList()
    }

    private class NamespaceColumn : ColumnInfo<TwigPath, String>("Namespace") {
        override fun valueOf(item: TwigPath): String? = item.namespace
    }

    private class PathColumn(private val project: Project) : ColumnInfo<TwigPath, String>("Path") {
        override fun valueOf(item: TwigPath): String? = item.getRelativePath(project)
    }

    private class CustomColumn : ColumnInfo<TwigPath, String>("Parser") {
        override fun valueOf(item: TwigPath): String? = if (item.isCustomPath) "Custom" else "Internal"
    }

    private class TypeColumn : ColumnInfo<TwigPath, String>("Type") {
        override fun valueOf(item: TwigPath): String? = item.namespaceType.toString()
    }

    private abstract class BooleanColumn(name: String) : ColumnInfo<TwigPath, Boolean>(name) {
        override fun isCellEditable(item: TwigPath): Boolean = true

        override fun getColumnClass(): Class<*> = Boolean::class.javaObjectType
    }

    private inner class DisableColumn : BooleanColumn("on") {
        override fun valueOf(item: TwigPath): Boolean = item.isEnabled

        override fun setValue(item: TwigPath, value: Boolean) {
            val index = tableView.listTableModel.items.indexOf(item)
            if (index >= 0) {
                val newTwigPath = TwigPath.createClone(item, value)
                tableView.listTableModel.removeRow(index)
                tableView.listTableModel.insertRow(index, newTwigPath)
            }
        }

        override fun getWidth(table: JTable): Int = 50
    }

    private fun openTwigPathDialog(twigPath: TwigPath?) {
        val dialog = if (twigPath == null) {
            TwigNamespaceDialog(project, tableView)
        } else {
            TwigNamespaceDialog(project, tableView, twigPath)
        }

        val dim = Dimension()
        dim.setSize(500, 190)
        dialog.title = "Twig Namespace"
        dialog.minimumSize = dim
        dialog.pack()
        dialog.setLocationRelativeTo(panel1)

        dialog.isVisible = true
    }
}
